src/NumberTheory/vinberg.jl

```diff
@@ -0,0 +1,322 @@
+    end
+    @req stopping_condition < length(Eigenvalues) "Q is not of signature (1, n)"
+    return _rescale_primitive(ZZ.(eigenspace(Q, x)))
+  else
+    @req (v0*Q*transpose(v0))[1, 1] > 0 "v0 has non positive inner product"
+    @req isone(reduce(gcd, v)) "v0 is not primitive"
+    return v0
+  end
+end
+
+@doc raw"""
+    _all_root_lengths(Q::ZZMatrix) -> Vector{ZZRingElem}
+
+If the user does not want any specific root lengths, we take all of them.
+
+For `Q` the matrix representing the hyperbolic reflection lattice and `l` a length to check,
+it is necessary that `l` divides $2.i$ for `l` being a possible root length,  
+with `i` being the level of `Q`, i.e. the last invariant (biggest entry) of the Smith normal form of `Q`.
+Prf: `r` being a root implies that $\frac{2.r.L}{r^2}$ is a subset of $\Z$
+     `r` being primitive implies that $2.i.\Z$ is a subset of $2.r.L$,
+     which is a subset of $(r^2).\Z$, what implies that $r^2$ divides $2.i$
+"""
+function _all_root_lengths(Q::ZZMatrix)
+  l = nrows(Q)
+  S = snf(Q)
+  return (-1) * divisors(2 * S[l, l])
+end
+
+@doc raw"""
+    _check_root_lengths(Q::ZZMatrix, root_lengths::Vector{ZZRingElem}) -> Vector{ZZRingElem}
+
+Check whether the given lengths are possible root lengths.
+Unnecessary roots are sorted out and reported if this is wanted.
+Return only the possible root lengths.
+"""
+function _check_root_lengths(Q::ZZMatrix, root_lengths::Vector{ZZRingElem})
+  l = nrows(Q)
+  S = snf(Q)
+  possible_root_lengths = divisors(2 * S[l, l])
+  if isempty(root_lengths)
+    return possible_root_lengths
+  end
+  result = ZZRingElem[]
+  for r in root_lengths
+    if abs(r) in possible_root_lengths
+      push!(result, r)
+    else
+      @vprintln :Vinberg 1 "$r is not a possible root length"
+    end
+  end
+  @req !isempty(result) "No possible root lengths found"
+  return result
+end
+
+@doc raw"""
+    _distance_indices(upper_bound, root_lengths::Vector{ZZRingElem}) -> Vector{Tuple{Int, ZZRingElem}}
+
+Returns all possible tuples $(n, l)$ with `n` a natural number and `l` contained in `root_lengths`,
+sorted by increase of the value $\frac{n^2}{l}$, stopping by `upper_bound`.
+"""
+function _distance_indices(upper_bound, root_lengths::Vector{ZZRingElem})
+
+  result = Tuple{Int,ZZRingElem}[]
+
+  for l in root_lengths
+    x0 = sqrt(big(abs(l * upper_bound))) # consider n^2//l < upper_bound => n < sqrt(l*upper_bound)
+    x = isqrt(x0 * upper_bound) 
+    for n in 1:x
+      push!(result, (n, l))
+    end
+  end
+
+  sort!(result; by=x -> (x[1]^2) // abs(x[2]))
+  return result
+end
+
+@doc raw"""
+    _rescale_primitive(v::ZZMatrix) -> ZZMatrix
+
+Check if `v` is primitive, which is equivalent to checking whether the greatest common divisor of the entries of `v`
+is equal to 1. If `v` is not primitive, we divide `v` by its gcd.
+"""
+function _rescale_primitive(v::ZZMatrix)
+  gcd = reduce(gcd, v)
+  if isone(gcd)
+    return v
+  else
+    return (1 // gcd * v)
+  end
```

Comment:
Minor comment: I would recommend not re-using a name like `gcd` for a local variable, perhaps call it `g`.

More importantlu, though, I don't think the `else` part works, or at least it doesn't when I tried it locally:
```
julia> v = matrix(ZZ, 6 * [1 2 ; 3 4])
[ 6   12]
[18   24]

julia> g = reduce(gcd, v)
6

julia> (1 // g * v)
ERROR: Cannot promote to common type
```
Even if it works, I would expect this to return a `QQMatrix`, which is probably not what you want.  However, this works:
```
julia> v / g
[1   2]
[3   4]
```

So overall perhaps use this?
```suggestion
  g = reduce(gcd, v)
  if isone(g)
    return v
  else
    return v / g
  end
```
or even
```suggestion
  g = reduce(gcd, v)
  return isone(g) ? v : v / g
```